File: controller/emulator/machine.py
# -*- coding: utf-8 -*-
# Машина — MK-52 emulator orchestrator. Ported from emulator/emulator.js
# (Такт, Шаг, Нажатие_кнопки, Отобразить_индикатор, Включить, Выключить).
# Wires four chips (ИК1302, ИК1303, ИР2_1, ИР2_2) and drives them from a
# background thread at the same 30 ms tick rate as the original setInterval.

import os
import threading
import time

from .chips import ИК13, ИР2
from .loader import Адрес_команды, ПРОГРАММНЫХ_ШАГОВ, разобрать_команду
from .rom import ПЗУ

Символы_разрядов = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "-", "L", "С", "Г", "Е", " "]


class Машина:

    ПЕРИОД_ШАГА = 0.030  # seconds — matches setInterval(Шаг, 30) in JS
    # Шаг iteration count. Original JS used 560 (calibrated so chip-time matches
    # real-time). CPython spends 89% of Такт runtime on int bit-ops; at 560 a
    # Шаг takes ~76 ms wall-clock on a Mac, so we run at ИТЕРАЦИЙ_В_ШАГЕ/560 of
    # original. Under PyPy the JIT makes Шаг ~65× faster — 560 fits well inside
    # the 30 ms tick. On a Pi Zero (armv6, CPython only, no PyPy build) Šaг at
    # 200 iters takes ~1270 ms wall-clock and blocks press_button that long; a
    # value like 10–20 keeps the chip lock cycling fast so keystrokes feel
    # responsive, at the same effective chip rate (limited by host CPU anyway).
    # Override via MK52_ITERS_PER_SHAG env var. Default 200 suits a desktop.
    ИТЕРАЦИЙ_В_ШАГЕ = int(os.environ.get("MK52_ITERS_PER_SHAG", "200"))

    def __init__(self, on_display=None, on_log=None):
        self.on_display = on_display
        self.on_log = on_log
        self._lock = threading.Lock()
        self._stop = threading.Event()
        self._thread = None
        self.Мера_угла = 10  # 10=радианы, 11=градусы, 12=грады

        self.ИК1302 = ИК13()
        self.ИК1303 = ИК13()
        self.ИР2_1 = ИР2()
        self.ИР2_2 = ИР2()
        self._Заполнить_ПЗУ()

        self.Индикатор = [0] * 13
        self.Индик_зпт = [False] * 13
        self.Индик_ст = [-1] * 13

    def _Заполнить_ПЗУ(self):
        self.ИК1302.ПЗУ_команд = ПЗУ["ИК1302"]["команды"]
        self.ИК1302.ПЗУ_синхропрограмм = ПЗУ["ИК1302"]["синхропрограммы"]
        self.ИК1302.ПЗУ_микрокоманд = ПЗУ["ИК1302"]["микрокоманды"]
        self.ИК1303.ПЗУ_команд = ПЗУ["ИК1303"]["команды"]
        self.ИК1303.ПЗУ_синхропрограмм = ПЗУ["ИК1303"]["синхропрограммы"]
        self.ИК1303.ПЗУ_микрокоманд = ПЗУ["ИК1303"]["микрокоманды"]

    def Такт(self):
        # Reference implementation, kept for clarity; the hot path in Шаг
        # inlines this to skip method-dispatch overhead in CPython.
        self.ИК1302.вход = self.ИР2_2.выход
        self.ИК1302.Такт()
        self.ИК1303.вход = self.ИК1302.выход
        self.ИК1303.Такт()
        self.ИР2_1.вход = self.ИК1303.выход
        self.ИР2_1.Такт()
        self.ИР2_2.вход = self.ИР2_1.выход
        self.ИР2_2.Такт()
        self.ИК1302.M[((self.ИК1302.микротакт >> 2) + 41) % 42] = self.ИР2_2.выход

    def Шаг(self):
        # Pre-bind everything used in the inner loop. CPython turns each
        # self.X / obj.method into a dict lookup; binding once cuts ~20% off
        # the 23 520-iteration hot path.
        ИК1302 = self.ИК1302
        ИК1303 = self.ИК1303
        ИР2_1 = self.ИР2_1
        ИР2_2 = self.ИР2_2
        ИК1302_Такт = ИК1302.Такт
        ИК1303_Такт = ИК1303.Такт
        ИК1302_M = ИК1302.M
        ИР2_1_M = ИР2_1.M
        ИР2_2_M = ИР2_2.M
        Индикатор = self.Индикатор
        Индик_зпт = self.Индик_зпт

        ИК1303.клав_y = 1
        ИК1303.клав_x = self.Мера_угла
        ИК1302_R = ИК1302.R
        ИК1302_зпт_внутр = ИК1302.индик_зпт

        # ИР2 state read once; written back after the hot loop.
        ИР2_1_мт = ИР2_1.микротакт
        ИР2_2_мт = ИР2_2.микротакт
        ИР2_2_выход = ИР2_2.выход

        for _ in range(self.ИТЕРАЦИЙ_В_ШАГЕ):
            for _ in range(42):
                ИК1302.вход = ИР2_2_выход
                ИК1302_Такт()
                ИК1303.вход = ИК1302.выход
                ИК1303_Такт()
                # ИР2_1.Такт() inlined: shift one slot of M[252].
                ИР2_1_вход_byte = ИК1303.выход
                ИР2_1_выход = ИР2_1_M[ИР2_1_мт]
                ИР2_1_M[ИР2_1_мт] = ИР2_1_вход_byte
                ИР2_1_мт = ИР2_1_мт + 1
                if ИР2_1_мт > 251:
                    ИР2_1_мт = 0
                # ИР2_2.Такт() inlined.
                ИР2_2_выход = ИР2_2_M[ИР2_2_мт]
                ИР2_2_M[ИР2_2_мт] = ИР2_1_выход
                ИР2_2_мт = ИР2_2_мт + 1
                if ИР2_2_мт > 251:
                    ИР2_2_мт = 0
                # ИК1302 feedback path (same as Такт()).
                _i = (ИК1302.микротакт >> 2) + 41
                if _i >= 42:
                    _i -= 42
                ИК1302_M[_i] = ИР2_2_выход
            if ИК1302.обн_индик:
                Индикатор[0] = ИК1302_R[24]
                Индикатор[1] = ИК1302_R[21]
                Индикатор[2] = ИК1302_R[18]
                Индикатор[3] = ИК1302_R[15]
                Индикатор[4] = ИК1302_R[12]
                Индикатор[5] = ИК1302_R[9]
                Индикатор[6] = ИК1302_R[6]
                Индикатор[7] = ИК1302_R[3]
                Индикатор[8] = ИК1302_R[0]
                Индикатор[9] = ИК1302_R[33]
                Индикатор[10] = ИК1302_R[30]
                Индикатор[11] = ИК1302_R[27]
                Индик_зпт[0] = ИК1302_зпт_внутр[9]
                Индик_зпт[1] = ИК1302_зпт_внутр[8]
                Индик_зпт[2] = ИК1302_зпт_внутр[7]
                Индик_зпт[3] = ИК1302_зпт_внутр[6]
                Индик_зпт[4] = ИК1302_зпт_внутр[5]
                Индик_зпт[5] = ИК1302_зпт_внутр[4]
                Индик_зпт[6] = ИК1302_зпт_внутр[3]
                Индик_зпт[7] = ИК1302_зпт_внутр[2]
                Индик_зпт[8] = ИК1302_зпт_внутр[1]
                Индик_зпт[9] = ИК1302_зпт_внутр[12]
                Индик_зпт[10] = ИК1302_зпт_внутр[11]
                Индик_зпт[11] = ИК1302_зпт_внутр[10]
                ИК1302.обн_индик = False

        # Flush ИР2 state back.
        ИР2_1.микротакт = ИР2_1_мт
        ИР2_2.микротакт = ИР2_2_мт
        ИР2_2.выход = ИР2_2_выход

        Индик_ст = self.Индик_ст
        обновить = False
        for i in range(13):
            if Индик_ст[i] != Индикатор[i]:
                обновить = True
            Индик_ст[i] = Индикатор[i]
        if обновить:
            self.Отобразить_индикатор()
        ИК1302.клав_x = 0
        ИК1302.клав_y = 0

    def Отобразить_индикатор(self):
        is_dimmed = self.ИК1302.запятая != 11
        digits = ""
        points = ""
        for i in range(12):
            digits += Символы_разрядов[self.Индикатор[i]]
            points += ("," if self.Индик_зпт[i] else " ")
        if self.on_display is not None:
            self.on_display(digits, points, is_dimmed)

    def Нажатие_кнопки(self, x, y):
        # Run two Šaги per press: the first with the key set (chip's keyboard
        # scan registers it), the second with the key cleared (chip's post-
        # press microcode settles: exit digit-entry after В↑, finalize push,
        # etc.). On fast hosts both fit inside a single press_button so the
        # cost is trivial; on slow hosts (Pi Zero) skipping the second Šaг
        # leaves the chip mid-settle when the next press arrives and merges
        # state across keys (e.g. "7 В↑ 3 +" produces 14 instead of 10).
        with self._lock:
            self.ИК1302.клав_x = x
            self.ИК1302.клав_y = y
            self.Шаг()
            self.Шаг()

    # Compatibility alias matching the old Emulator.press_button.
    press_button = Нажатие_кнопки

    def Ввести_код(self, source: str) -> int:
        """Parse MK-52 source and write opcodes directly into program memory.
        Returns the number of program steps written. Tokens are whitespace-
        separated; an optional 'NN.' line-number prefix is stripped.
        """
        tokens = source.split()
        # Strip optional 'NN.' / 'AN.' line-number prefixes; the JS parser
        # used the prefix to set the target step index, but for a simple
        # in-order listing we just drop it.
        стрипнутые = []
        for t in tokens:
            if len(t) >= 3 and t[2] == "." and (t[0].isdigit() or t[0] in "A-") and t[1].isdigit():
                t = t[3:]
            if t:
                стрипнутые.append(t)
        commands = стрипнутые[:ПРОГРАММНЫХ_ШАГОВ]

        chips = {
            1: self.ИР2_1, 2: self.ИР2_2,
            3: self.ИК1302, 4: self.ИК1303,
        }
        with self._lock:
            перестановка = (self.ИР2_1.микротакт // 84) % len(
                # 3 permutations
                (0, 1, 2)
            )
            def записать(номер, код):
                ст = код >> 4
                мл = код & 0xF
                chip_id, адрес = Адрес_команды(номер, перестановка)
                chip = chips[chip_id]
                chip.M[адрес] = ст
                chip.M[адрес - 3] = мл

            for i, tok in enumerate(commands):
                записать(i, разобрать_команду(tok))
            # Zero remaining program steps so a previous listing doesn't bleed through.
            for i in range(len(commands), ПРОГРАММНЫХ_ШАГОВ):
                записать(i, 0)
            # Ensure the program counter is reset (В/О behavior).
            self.ИК1302.R[36] = 0
            self.ИК1302.R[39] = 0
        return len(commands)

    def _Цикл(self):
        next_at = time.monotonic()
        while not self._stop.is_set():
            with self._lock:
                self.Шаг()
            next_at += self.ПЕРИОД_ШАГА
            sleep = next_at - time.monotonic()
            if sleep > 0:
                self._stop.wait(sleep)
            else:
                next_at = time.monotonic()  # we fell behind; resync

    def __enter__(self):
        if self.on_log is not None:
            self.on_log("Initializing...")
        # Prime the chips with one Шаг so the first display refresh fires.
        with self._lock:
            self.Шаг()
        self._stop.clear()
        self._thread = threading.Thread(target=self._Цикл, daemon=True, name="МК-52 Шаг")
        self._thread.start()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self._stop.set()
        if self._thread is not None:
            self._thread.join(timeout=1.0)
            self._thread = None

File: go/mk52/machine.go
// Машина — MK-52 emulator orchestrator. Ported from controller/emulator/machine.py.
// Wires four chips (ИК1302, ИК1303, ИР2_1, ИР2_2) and drives them from a
// background goroutine at the same 30 ms tick rate as the JS original.

package mk52

import (
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

const ПериодШага = 30 * time.Millisecond

// ИТЕРАЦИЙ_В_ШАГЕ — inner-loop iteration count. Original JS used 560 (calibrated
// so chip-time matches real-time). Override via MK52_ITERS_PER_SHAG.
var ИТЕРАЦИЙ_В_ШАГЕ = envInt("MK52_ITERS_PER_SHAG", 200)

// СимволыРазрядов maps display nibble (0-15) to displayed character.
var СимволыРазрядов = []string{
	"0", "1", "2", "3", "4", "5", "6", "7", "8", "9",
	"-", "L", "С", "Г", "Е", " ",
}

// DisplayFn receives the formatted display state from the chip every Šaг
// when content changes.
type DisplayFn func(digits, points string, isDimmed bool)

// LogFn receives status strings (init/log).
type LogFn func(text string)

type Машина struct {
	OnDisplay DisplayFn
	OnLog     LogFn
	МераУгла  int // 10=radians, 11=degrees, 12=grads

	mu sync.Mutex

	ИК1302 *ИК13
	ИК1303 *ИК13
	ИР2_1  *ИР2
	ИР2_2  *ИР2

	Индикатор [13]uint8
	ИндикЗпт  [13]bool
	индикСт   [13]int

	stop chan struct{}
	done chan struct{}
}

func envInt(name string, def int) int {
	v := os.Getenv(name)
	if v == "" {
		return def
	}
	n, err := strconv.Atoi(v)
	if err != nil {
		return def
	}
	return n
}

func NewМашина() *Машина {
	m := &Машина{
		МераУгла: 10,
		ИК1302: &ИК13{
			ПЗУ_команд:         ПЗУ_ИК1302_команды,
			ПЗУ_синхропрограмм: ПЗУ_ИК1302_синхропрограммы,
			ПЗУ_микрокоманд:    ПЗУ_ИК1302_микрокоманды,
		},
		ИК1303: &ИК13{
			ПЗУ_команд:         ПЗУ_ИК1303_команды,
			ПЗУ_синхропрограмм: ПЗУ_ИК1303_синхропрограммы,
			ПЗУ_микрокоманд:    ПЗУ_ИК1303_микрокоманды,
		},
		ИР2_1: &ИР2{},
		ИР2_2: &ИР2{},
	}
	for i := range m.индикСт {
		m.индикСт[i] = -1
	}
	return m
}

// Шаг runs one chip step: ИТЕРАЦИЙ_В_ШАГЕ iterations × 42 microticks each,
// driving all four chips together. Hot path; ported from machine.py Шаг().
func (m *Машина) Шаг() {
	ИК1302 := m.ИК1302
	ИК1303 := m.ИК1303
	ИР2_1 := m.ИР2_1
	ИР2_2 := m.ИР2_2

	ИК1303.клав_y = 1
	ИК1303.клав_x = m.МераУгла

	// Pre-pull state into locals for hot loop.
	ИР2_1_мт := ИР2_1.микротакт
	ИР2_2_мт := ИР2_2.микротакт
	ИР2_2_выход := ИР2_2.выход

	for iter := 0; iter < ИТЕРАЦИЙ_В_ШАГЕ; iter++ {
		for inner := 0; inner < 42; inner++ {
			// Inlined ИК1302.Такт / ИК1303.Такт via their existing methods.
			ИК1302.вход = ИР2_2_выход
			ИК1302.Такт()
			ИК1303.вход = ИК1302.выход
			ИК1303.Такт()

			// Inlined ИР2_1.Такт: shift one slot in M[252].
			ИР2_1_вход := ИК1303.выход
			ИР2_1_выход := ИР2_1.M[ИР2_1_мт]
			ИР2_1.M[ИР2_1_мт] = ИР2_1_вход
			ИР2_1_мт++
			if ИР2_1_мт > 251 {
				ИР2_1_мт = 0
			}

			// Inlined ИР2_2.Такт.
			ИР2_2_выход = ИР2_2.M[ИР2_2_мт]
			ИР2_2.M[ИР2_2_мт] = ИР2_1_выход
			ИР2_2_мт++
			if ИР2_2_мт > 251 {
				ИР2_2_мт = 0
			}

			// ИК1302 feedback path: M[((микротакт >> 2) + 41) mod 42] ← ИР2_2.выход
			i := (ИК1302.микротакт >> 2) + 41
			if i >= 42 {
				i -= 42
			}
			ИК1302.M[i] = ИР2_2_выход
		}

		if ИК1302.обн_индик {
			R := &ИК1302.R
			зпт := &ИК1302.индик_зпт
			m.Индикатор[0] = R[24]
			m.Индикатор[1] = R[21]
			m.Индикатор[2] = R[18]
			m.Индикатор[3] = R[15]
			m.Индикатор[4] = R[12]
			m.Индикатор[5] = R[9]
			m.Индикатор[6] = R[6]
			m.Индикатор[7] = R[3]
			m.Индикатор[8] = R[0]
			m.Индикатор[9] = R[33]
			m.Индикатор[10] = R[30]
			m.Индикатор[11] = R[27]
			m.ИндикЗпт[0] = зпт[9]
			m.ИндикЗпт[1] = зпт[8]
			m.ИндикЗпт[2] = зпт[7]
			m.ИндикЗпт[3] = зпт[6]
			m.ИндикЗпт[4] = зпт[5]
			m.ИндикЗпт[5] = зпт[4]
			m.ИндикЗпт[6] = зпт[3]
			m.ИндикЗпт[7] = зпт[2]
			m.ИндикЗпт[8] = зпт[1]
			m.ИндикЗпт[9] = зпт[12]
			m.ИндикЗпт[10] = зпт[11]
			m.ИндикЗпт[11] = зпт[10]
			ИК1302.обн_индик = false
		}
	}

	// Flush ИР2 microtakt state back.
	ИР2_1.микротакт = ИР2_1_мт
	ИР2_2.микротакт = ИР2_2_мт
	ИР2_2.выход = ИР2_2_выход

	// Check whether the display changed; fire on_display if so.
	обновить := false
	for i := 0; i < 13; i++ {
		if m.индикСт[i] != int(m.Индикатор[i]) {
			обновить = true
		}
		m.индикСт[i] = int(m.Индикатор[i])
	}
	if обновить {
		m.ОтобразитьИндикатор()
	}
	ИК1302.клав_x = 0
	ИК1302.клав_y = 0
}

// ОтобразитьИндикатор formats the display state and invokes OnDisplay.
func (m *Машина) ОтобразитьИндикатор() {
	isDimmed := m.ИК1302.запятая != 11
	var digits, points strings.Builder
	for i := 0; i < 12; i++ {
		digits.WriteString(СимволыРазрядов[m.Индикатор[i]])
		if m.ИндикЗпт[i] {
			points.WriteByte(',')
		} else {
			points.WriteByte(' ')
		}
	}
	if m.OnDisplay != nil {
		m.OnDisplay(digits.String(), points.String(), isDimmed)
	}
}

// НажатиеКнопки presses key at (x, y), runs two Šaги (press + settle).
func (m *Машина) НажатиеКнопки(x, y int) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.ИК1302.клав_x = x
	m.ИК1302.клав_y = y
	m.Шаг()
	m.Шаг()
}

// ВвестиКод parses MK-52 source and writes opcodes directly into program
// memory. Returns the number of program steps written. Tokens are whitespace-
// separated; an optional "NN." line-number prefix is stripped.
func (m *Машина) ВвестиКод(source string) (int, error) {
	tokens := strings.Fields(source)
	cleaned := make([]string, 0, len(tokens))
	for _, t := range tokens {
		if len(t) >= 3 && t[2] == '.' &&
			((t[0] >= '0' && t[0] <= '9') || t[0] == 'A' || t[0] == '-') &&
			t[1] >= '0' && t[1] <= '9' {
			t = t[3:]
		}
		if t != "" {
			cleaned = append(cleaned, t)
		}
	}
	if len(cleaned) > ПРОГРАММНЫХ_ШАГОВ {
		cleaned = cleaned[:ПРОГРАММНЫХ_ШАГОВ]
	}

	m.mu.Lock()
	defer m.mu.Unlock()
	перестановка := (m.ИР2_1.микротакт / 84) % 3
	chips := [5]interface {
		writeM(addr int, val uint8)
	}{nil, m.ИР2_1, m.ИР2_2, m.ИК1302, m.ИК1303}

	записать := func(номер int, код byte) {
		ст := uint8(код >> 4)
		мл := uint8(код & 0xf)
		chipID, адрес := АдресКоманды(номер, перестановка)
		chips[chipID].writeM(адрес, ст)
		chips[chipID].writeM(адрес-3, мл)
	}

	for i, tok := range cleaned {
		op, err := РазобратьКоманду(tok)
		if err != nil {
			return 0, err
		}
		записать(i, op)
	}
	for i := len(cleaned); i < ПРОГРАММНЫХ_ШАГОВ; i++ {
		записать(i, 0)
	}
	m.ИК1302.R[36] = 0
	m.ИК1302.R[39] = 0
	return len(cleaned), nil
}

// writeM helpers so ИР2 and ИК13 can be addressed uniformly from ВвестиКод.
func (chip *ИР2) writeM(addr int, val uint8)  { chip.M[addr] = val }
func (chip *ИК13) writeM(addr int, val uint8) { chip.M[addr] = val }

// Start spawns the background ticker goroutine. Stop() to halt.
func (m *Машина) Start() {
	if m.OnLog != nil {
		m.OnLog("Initializing...")
	}
	m.mu.Lock()
	m.Шаг() // prime so first display refresh fires
	m.mu.Unlock()
	m.stop = make(chan struct{})
	m.done = make(chan struct{})
	go m.цикл()
}

// Stop signals the ticker to exit and waits for it.
func (m *Машина) Stop() {
	if m.stop != nil {
		close(m.stop)
		<-m.done
		m.stop = nil
	}
}

func (m *Машина) цикл() {
	defer close(m.done)
	nextAt := time.Now()
	for {
		select {
		case <-m.stop:
			return
		default:
		}
		m.mu.Lock()
		m.Шаг()
		m.mu.Unlock()
		nextAt = nextAt.Add(ПериодШага)
		sleep := time.Until(nextAt)
		if sleep > 0 {
			select {
			case <-m.stop:
				return
			case <-time.After(sleep):
			}
		} else {
			nextAt = time.Now() // fell behind; resync
		}
	}
}
